### ISSUE TITLE
Any support for 8BitDo SN30 Pro USB?

### ISSUE THREAD

**Issue by mconigliaro** [360Controller/360Controller#964]
### Type of Controller
8BitDo SN30 Pro USB

### OS Version
10.13.1

### Driver Version
- 0.16.11
- 1.0.0-alpha.5

### Connection Method
Wired

### Device Name and Info
```
Controller:

  Product ID:	0x028e
  Vendor ID:	0x045e  (Microsoft Corporation)
  Version:	1.14
  Serial Number:	Controller
  Speed:	Up to 12 Mb/sec
  Manufacturer:	Controller
  Location ID:	0x14b00000 / 5
  Current Available (mA):	500
  Extra Operating Current (mA):	0
```

### Response of Controller
Vibrates and green LED turns on

### Response in Preference Pane
"No devices found"

### Detailed Explaination of Issue
I'm trying to use this controller for OpenEMU, but I can't figure out how to get it recognized. I've already upgraded the firmware on the controller to v1.03. On the 8BitDo website, there's a question in the FAQ: https://support.8bitdo.com/faq/sn30-pro-usb-controller.html

```
Can I use it on macOS?

Yes, you can. But you need to get X-input drive first onto your computer.
```

There’s also a question here: https://www.amazon.com/dp/B07R481CVY

```
Question: Will it work on Mac?
                                                                                    
Yes, it will, you just need the 360Controller driver installed. Here is the link:

https://github.com/360Controller/360Controller
```

I’ve tried installing the 360Controller package (via `brew install xbox360-controller-driver-unofficial` and by manually installing the latest version 1.0.0-alpha.5) but the controller still isn’t recognized. Is there something else I need to do? Is this controller supported at all?

**Comment by FranticRain** [comment]
A couple of things here.
1. Never use that `homebrew` install of this driver. I don't know who supports it, but it's only usable for super power users who know EXACTLY what they are using by using it. I hate that it exists.
2. Since the device you are using emulates an Xbox 360 controller, including stealing the vendor and product ID of the Xbox 360 controller, it is not supported by this driver. We do not support devices that do not follow the USB specification by stealing vendor and product IDs.
3. I can help you with getting the driver working on your machine, which appears to be the real problem here. Since you're running 10.13, you'll need to use either the stable or alpha 3 version.

I'll leave this issue open for a bit for any further help you might need regarding getting the driver working, but support for that controller is out of scope.

**Comment by mconigliaro** [comment]
If I'm understanding correctly, this controller *should* just work given that it's emulating an Xbox 360 controller, but you're not going to support it officially. Fair enough.

FWIW, I've uninstalled `xbox360-controller-driver-unofficial` via homebrew, uninstalled alpha 5 via the advanced tab of the preference window, and manually installed alpha 3, but I'm not seeing any different behavior. I just get "No devices found" in the dropdown list.

**Comment by FranticRain** [comment]
Yes. You've understood correctly.

Could you run this in your terminal to try to manually load the driver and check if it is working as expected? Let me know what it says.
```
sudo kextutil /Library/Extensions/360Controller.kext
```

**Comment by mconigliaro** [comment]
```
$ sudo kextutil /Library/Extensions/360Controller.kext
Password:
Notice: /Library/Extensions/360Controller.kext has debug properties set.
Warnings:
    Personality CFBundleIdentifier differs from containing kext's (not necessarily a mistake, but rarely done):
        ChatPadKeyboardEvents

(kernel) kxld[com.mice.driver.Xbox360Controller]: The super class vtable '__ZTV11IOHIDDevice' for vtable '__ZTV20ChatPadKeyboardClass' is out of date. Make sure your kext has been built against the correct headers.
(kernel) kxld[com.mice.driver.Xbox360Controller]: The super class vtable '__ZTV11IOHIDDevice' for vtable '__ZTV22Xbox360ControllerClass' is out of date. Make sure your kext has been built against the correct headers.
(kernel) kxld[com.mice.driver.Xbox360Controller]: The super class vtable '__ZTV11IOHIDDevice' for vtable '__ZTV20ChatPadKeyboardClass' is out of date. Make sure your kext has been built against the correct headers.
(kernel) kxld[com.mice.driver.Xbox360Controller]: The super class vtable '__ZTV11IOHIDDevice' for vtable '__ZTV22Xbox360ControllerClass' is out of date. Make sure your kext has been built against the correct headers.
(kernel) kxld[com.mice.driver.Xbox360Controller]: The __ZTV22Xbox360Pretend360Class was not patched because its parent, the __ZTV22Xbox360ControllerClass, was not patchable.
(kernel) kxld[com.mice.driver.Xbox360Controller]: The __ZTV22XboxOneControllerClass was not patched because its parent, the __ZTV22Xbox360ControllerClass, was not patchable.
(kernel) kxld[com.mice.driver.Xbox360Controller]: The __ZTV22XboxOnePretend360Class was not patched because its parent, the __ZTV22XboxOneControllerClass, was not patchable.
(kernel) kxld[com.mice.driver.Xbox360Controller]: The __ZTV27XboxOriginalControllerClass was not patched because its parent, the __ZTV22Xbox360ControllerClass, was not patchable.
(kernel) Can't load kext com.mice.driver.Xbox360Controller - link failed.
(kernel) Failed to load executable for kext com.mice.driver.Xbox360Controller.
(kernel) Kext com.mice.driver.Xbox360Controller failed to load (0xdc008016).
(kernel) Failed to load kext com.mice.driver.Xbox360Controller (error 0xdc008016).
Failed to load /Library/StagedExtensions/Library/Extensions/360Controller.kext - (libkern/kext) link error.
Check library declarations for your kext with kextlibs(8).
```

Wild guess, but maybe 10.13.1 is too old?

**Comment by FranticRain** [comment]
It's not too old, it's just that a change was made at some point to where the driver was being built against macOS 10.14 instead of 10.11. That meant that, unbeknownst to me, the driver is broken on all versions of macOS before 10.14. Now what I'm really curious about is the fact that alpha 3 doesn't work. I may need to check my facts, but I'm pretty sure that alpha 3 shipped with a system that had an individual version compiled for each version of macOS 10.11 and after. So one built for 10.11, one built for 10.12, etc. So I *thought* that that would fix it. Can you do the same test against the 0.16.11 version of the driver? Sorry to use you as a guinea pig like this, but I get yelled at by issue creators for not **immediately** supporting beta versions of macOS, so I don't get a lot of opportunity to test against older versions of macOS. I'm already dual booting 10.14/10.15 just so I can build the dang driver. I'm not sure this machine can handle another older version.

To put it simply, this is a really hard problem to solve, because I have to build the driver against the oldest supported version in order to have backwards compatibility. But macOS has changed some things between now and then that need to be addressed, but can't be built on older versions of macOS. It's a mess. I thought I had a solution, but clearly it didn't work.

**Comment by FranticRain** [comment]
Okay I did a quick and dirty test and removed a bunch of stuff to just get it to work on 10.13. It's not a permanent solution, but this *should* work.

[360ControllerInstall.dmg.zip](https://github.com/360Controller/360Controller/files/3812645/360ControllerInstall.dmg.zip)

**Comment by mconigliaro** [comment]
No problem man. I appreciate the help. If I can't get this controller working, I guess I'll have to send it back and try something else.

So here's what happens with the 0.16.11 version of the driver:

```
$ sudo kextutil /Library/Extensions/360Controller.kext
Password:
Notice: /Library/Extensions/360Controller.kext has debug properties set.
Warnings:
    Personality CFBundleIdentifier differs from containing kext's (not necessarily a mistake, but rarely done):
        ChatPadKeyboardEvents
```

And it looks like it works now! I had to close and reopen the preferences window to make the controller show up in the dropdown list (`Xbox 360 Wired Controller (Wired)`), but then I launched OpenEMU and was able to play a round of Tetris on NES. Thanks! Let me know if you still want me to try that `360ControllerInstall.dmg.zip` you just posted or if I can help with anything else.

**Comment by FranticRain** [comment]
I'd appreciate it if you'd give it a shot just to make sure. Glad to hear that you got stable working. I'm going to go ahead and close this down since you got it working. Feel free to respond with your findings.

**Comment by mconigliaro** [comment]
FYI, that one-off version you posted worked fine for me. Thanks again for the help!

**Comment by mconigliaro** [comment]
OK, I'm back after upgrading to Catalina (10.15.7). Interestingly, I wasn't able to get the controller recognized on `0.16.11`, but `1.0.0-alpha.6` seems to work.

**Comment by JeanLecat** [comment]
Hello ! I amm using a Mac M1 and try to connect a wired SN30 like you. However I can't install 360Controller. I have the following  message (installation error in French) 
<img width="365" alt="image" src="https://user-images.githubusercontent.com/91464476/176473750-dc71386c-48de-41f9-b1cd-9f83c6e0a9d1.png">
